Newton merge 3, part 1: Add newton simple nonlinear material

MFraters opened the pull request:
split up of #1653, part 1

bangerth commented:
Same as in #1684: @gassmoeller , @anne-glerum ?

tjhei commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,115 @@
+    using namespace dealii;
+
+    /**
+     * A material model based on a simple powerlaw rheology and
```
"power law"

tjhei commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,115 @@
+
+        virtual double reference_viscosity () const;
+
+        virtual double reference_density () const;
```
what is this used for? This is not part of the main interface, right?

tjhei commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,115 @@
+
+      private:
+
+        double reference_T;
```
document

tjhei commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,115 @@
+    class SimpleNonlinear : public MaterialModel::Interface<dim>, public ::aspect::SimulatorAccess<dim>
+    {
+      public:
+        std::vector<double> compute_volume_fractions( const std::vector<double> &compositional_fields) const;
```
make protected

tjhei commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,115 @@
+     * @ingroup MaterialModels
+     */
+    template <int dim>
+    class SimpleNonlinear : public MaterialModel::Interface<dim>, public ::aspect::SimulatorAccess<dim>
```
I would like to avoid using the word "simple", even if this has anything to do with the "simple" model.

MFraters replied:
What is the reason for that? I could call it power law, which would be more appropriate, but I would also like to make a the distinction that that it is practically the simplest form of a power law possible, and not a power law like diffussion and/or dislocation creep.

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+    {
+      //set up additional output for the derivatives
+      MaterialModelDerivatives<dim> *derivatives;
+      derivatives = out.template get_additional_output<MaterialModelDerivatives<dim> >();
```
just use one line and initialize immediately.

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+                      if (edot_ii_strict > min_strain_rate[c] * min_strain_rate[c] && composition_viscosities[c] < max_viscosity[c] && composition_viscosities[c] > min_viscosity[c])
+                        {
+                          //strictly speaking the derivative is this: 0.5 * ((1/stress_exponent)-1) * std::pow(2,2) * out.viscosities[i] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i])
+                          composition_viscosities_derivatives[c] = 2 * (stress_exponent_inv-1) * composition_viscosities[c] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i]);
```
please make all numbers doubles explicitly: 2.0 1.0, etc.

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+                }
+
+              out.viscosities[i] = Utilities::weighted_p_norm_average(volume_fractions, composition_viscosities, viscosity_averaging_p);
+              Assert(dealii::numbers::is_finite(out.viscosities[i]),ExcMessage ("Error: Averaged viscosity is not finite."));
```
space after comma

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+                              ExcMessage ("Averaged viscosity to strain-rate devrivative is not finite."));
+
+                  derivatives->viscosity_derivative_wrt_pressure[i]    = 0;
+                  Assert (dealii::numbers::is_finite(derivatives->viscosity_derivative_wrt_pressure[i]),
```
I don't understand. Remove the assert? Can you remove the spaces in front of "="?

MFraters replied:
You are right, that's weird, there should be not assert

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+          out.thermal_conductivities[i] = thermal_conductivities;
+          // Compressibility at the given positions.
+          // The compressibility is given as
+          // $\frac 1\rho \frac{\partial\rho}{\partial p}$.
```
out of date comment

MFraters replied:
so what is it now?

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+    {
+      prm.enter_subsection ("Compositional fields");
+      {
+        prm.declare_entry ("Number of fields", "0",
```
I am confused. This shouldn't be here.

MFraters replied:
I need it for my tests to work

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+    SimpleNonlinear<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      // Can't use this->n_compositional_fields(), because some
+      // tests never initialize the simulator, but uses the material
```
I think we should fix the tests instead.

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+      // model directly.
+      prm.enter_subsection ("Compositional fields");
+      {
+        n_fields = prm.get_integer ("Number of fields")+1;
```
why +1?

MFraters replied:
Because this material mode was set up based on a material model which worked with a background compositional field. But we still need to store values for that field

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(SimpleNonlinear,
+                                   "simple nonlinear",
+                                   " TODO: rewrite... An implementation of a viscous rheology including diffusion"
```
please rewrite

tjhei commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+                                   " \\exp\\left(\\frac{E_i^* + PV_i^*}{n_iRT}\\right)\\]"
+                                   " \n\n"
+                                   " where $d$ is grain size, $i$ corresponds to diffusion or dislocation creep,"
+                                   " $\\dot{\\varepsilon}$ is the square root of the second invariant of the"
```
missing spaces at the end of these lines.

tjhei commented on tests/simple_nonlinear.prm:
```diff
@@ -0,0 +1 @@
+set Additional shared libraries = tests/libsimple_nonlinear.so
```
add a comment that describes this test.

gassmoeller commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,115 @@
+
+        unsigned int n_fields;
+        // averaging parameters
+        double viscosity_averaging_p;
```
averaging parameters does not explain what this is

gassmoeller commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,115 @@
+        std::vector<double> prefactor;
+        std::vector<double> stress_exponent;
+
+        unsigned int n_fields;
```
if this simply stores the number of compositional fields, please remove it. It just duplicates information that is easily available from the SimulatorAccess class.

MFraters replied:
This info is not available for the tests

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+    template <int dim>
+    std::vector<double>
+    SimpleNonlinear<dim>::
+    compute_volume_fractions( const std::vector<double> &compositional_fields) const
```
I am pretty sure this function is now used in several material models. Please open an issue about unifying them in the Utilities namespace at a later time.

MFraters replied:
done: #1670 ;)
will be on my list after the newton is merged, or when I am tired of newton for a moment :)

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+              // this material model (a general powerlaw) we do not need to worry about how to
+              // distribute the strain-rate and stress over the processes.
+              std::vector<double> composition_viscosities(volume_fractions.size());
+              std::vector<SymmetricTensor<2,dim> > composition_viscosities_derivatives(volume_fractions.size());
```
You only need this if derivatives are not 0, please check this.

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+                    {
+                      if (edot_ii_strict > min_strain_rate[c] * min_strain_rate[c] && composition_viscosities[c] < max_viscosity[c] && composition_viscosities[c] > min_viscosity[c])
+                        {
+                          //strictly speaking the derivative is this: 0.5 * ((1/stress_exponent)-1) * std::pow(2,2) * out.viscosities[i] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i])
```
What do you mean by this comment? Is this a different way to write the line below? Or do you mean the code is wrong? Please clarify that in the comment or remove it. Or is it a question for us?

MFraters replied:
no, just saying that when you derive the equation you will get this formulation. But I simply it as the implemented equation. I have updated the comment

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+                  derivatives->viscosity_derivative_wrt_strain_rate[i] = Utilities::derivative_of_weighted_p_norm_average(out.viscosities[i],volume_fractions, composition_viscosities, composition_viscosities_derivatives, viscosity_averaging_p);
+
+                  for (int x = 0; x < dim; x++)
+                    for (int y = 0; y < dim; y++)
```
should x and y be `unsigned int`?

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+    reference_density () const
+    {
+      return densities[0];
+    }
```
Remove this function, it is not longer used

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+      prm.enter_subsection ("Compositional fields");
+      {
+        prm.declare_entry ("Number of fields", "0",
+                           Patterns::Integer (0),
```
Yes, this should already be declared in Simulator. I am surprised this did not generate exceptions on your machine

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+
+
+          // SimpleNonlinear creep parameters
+          prm.declare_entry ("Prefactor", "1e-37",
```
Can you give this a more descriptive name? "Viscosity prefactor" comes to mind :smirk:

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+          prm.declare_entry ("Viscosity averaging p", "-1",
+                             Patterns::Double(),
+                             "This is the p value in the generalized weighed average eqation: "
+                             " mean = \\frac{1}{k}(\\sum_{i=1}^k \\big(c_i \\eta_{\\text{eff}_i}^p)\\big)^{\\frac{1}{p}}. "
```
I am pretty sure you need to encapsulate this into `$ .... $` otherwise the next time Timo will update the manual he will see some interesting errors.

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+      prm.enter_subsection ("Compositional fields");
+      {
+        n_fields = prm.get_integer ("Number of fields")+1;
+      }
```
Also by now you can remove this whole enter_subsection/exit_subsection call, and instead use `this->n_compositional_fields()`. This is a workaround that is not longer necessary since several years, but it keeps reappearing due to copy-paste.

MFraters replied:
as stated above, it doesn't work for the tests.

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+          thermal_diffusivity =  Utilities::possibly_extend_from_1_to_N (Utilities::string_to_double(Utilities::split_string_list(prm.get("Thermal diffusivity"))),n_fields,"Thermal diffusivity");
+          heat_capacity =  Utilities::possibly_extend_from_1_to_N (Utilities::string_to_double(Utilities::split_string_list(prm.get("Heat capacity"))),n_fields,"Heat capacity");
+
+          // ---- Compositional parameters
```
Remove the dashes.

gassmoeller commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,371 @@
+          thermal_expansivities =  Utilities::possibly_extend_from_1_to_N (Utilities::string_to_double(Utilities::split_string_list(prm.get("Thermal expansivities"))),n_fields,"Thermal expansivities");
+
+          // Rheological parameters
+          prefactor =  Utilities::possibly_extend_from_1_to_N (Utilities::string_to_double(Utilities::split_string_list(prm.get("Prefactor"))),n_fields,"Prefactor");
```
`viscosity_prefactor` would be more descriptive

gassmoeller commented on tests/simple_nonlinear.cc:
```diff
@@ -0,0 +1,280 @@
+  in_base.pressure[4] = 2e12;
+
+  /**
+   * We cant take to slow values, because then the difference in the
```
replace too slow by something more descriptive. too small strain rates?

gassmoeller commented on tests/simple_nonlinear.cc:
```diff
@@ -0,0 +1,280 @@
+
+  /**
+   * We cant take to slow values, because then the difference in the
+   * visocisty will be too small for the double accuracy which stores
```
viscosity

gassmoeller commented on tests/simple_nonlinear.cc:
```diff
@@ -0,0 +1,280 @@
+// run this function by initializing a global variable by it
+int ii = f(-1000); // Testing min function
+int iz = f(-2); // Testing min function
+int ij = f(-1.5); // Testing generalized p norm mean with negative p
```
Update or remove all comments in this block

gassmoeller commented on tests/simple_nonlinear/screen-output:
```diff
@@ -0,0 +1,176 @@
+oneone 3: Finite difference = -1.31479e+37. Analytical derivative = -0
+   Error: The derivative of the viscosity to the strain rate is too different from the analitical value.
+oneone 4: Finite difference = -1.4336e+22. Analytical derivative = -0
+   Error: The derivative of the viscosity to the strain rate is too different from the analitical value.
```
So it looks like a whole lot of the tests fail by a large amount. Do they pass on your machine?

MFraters replied:
I think it has to do with the way finite difference and analytical derivatives work. I posted my interpretation in a response to @tjhei question in #1684

MFraters commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,115 @@
+        std::vector<double> min_strain_rate;
+        std::vector<double> min_viscosity;
+        std::vector<double> max_viscosity;
+        std::vector<double> veff_coefficient;
```
it was a old left over parameter, which isn't used any more. I removed it.

MFraters commented:
I think that I have addressed all the comments. There are three things which need to be awnsered:

1) what should we call this material model? I could call it power law, which would be more appropriate, but I would also like to make a the distinction that that it is practically the simplest form of a power law possible, and not a power law like diffussion and/or dislocation creep.

2) How do we get the tests to work without declaring and parsing number of composition fields.

3) what should we do with the test result of the case constructed case which fails. This case is not realistic, but it is an interesting one. I also put it into #1684, but I guess here it is a more central place to discuss it. Or should I make an issue?

I had written something on this in a email to wolfgang a long time ago:

In short the problem is that for all common cases I found that the finite difference and the analytical solution are exactly the same. But with a strain rate in which the diagonal entries are the same the analytical solution gives -0 and the finite difference gives results on the order of the other strain rates (~-1e+20).

Here is the relevant code (There is also a minimum to edot_ii in the real code.):

   const double edot_ii = 2 * std::sqrt(0.5*deviator(in.strain_rate[i])*deviator(in.strain_rate[i]));
   const double stress_exponent_inv = (1./stress_exponent[c]);
   composition_viscosities[c] = std::pow(prefactor[c],-stress_exponent_inv) * std::pow(edot_ii,stress_exponent_inv-1);
   composition_viscosities_derivatives[c] = 2 * (stress_exponent_inv-1) * composition_viscosities[c] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i]);

I think this happens for the following reason: In the analytical equation, the deviator make the diagonal then zero, and the analytical solution then also ends up with zero entries on the diagonal. The finite difference method first gives the original strain rate tensor to the material model, and the deviator makes the diagonal zero. But then the finite difference method perturbs only one of the diagonal entries, which make the diagonal entries not equal any more and as a consequence the diagonal of the deviator will not contain zero entries any more, and I think this lead to a viscosity change on the order of what normally happens.

Thoughts?

gassmoeller commented:
/run-tests

MFraters commented:
@gassmoeller, I changed what we have discussed. So I guess it will be good to go for now when the testers are happy.
